File: client/src/lib/validators/regexValidator.test.ts
/**
 * Testes para o validador de regex com proteção ReDoS
 */

import { describe, it, expect } from 'vitest';
import { 
  validateRegex, 
  getRegexValidationError, 
  isRegexSafe,
  MAX_PATTERN_LENGTH,
  MAX_CAPTURE_GROUPS 
} from './regexValidator';

describe('regexValidator', () => {
  describe('validateRegex', () => {
    describe('Patterns válidos', () => {
      it('deve aceitar pattern simples', () => {
        const result = validateRegex('^[a-z]+$');
        expect(result.valid).toBe(true);
        expect(result.errorMessage).toBeNull();
      });

      it('deve aceitar pattern de email', () => {
        const result = validateRegex('^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+$');
        expect(result.valid).toBe(true);
      });

      it('deve aceitar pattern de telefone', () => {
        const result = validateRegex('^\\+?[0-9]{10,14}$');
        expect(result.valid).toBe(true);
      });

      it('deve aceitar pattern de CPF', () => {
        const result = validateRegex('^\\d{3}\\.\\d{3}\\.\\d{3}-\\d{2}$');
        expect(result.valid).toBe(true);
      });

      it('deve aceitar pattern de CEP', () => {
        const result = validateRegex('^\\d{5}-?\\d{3}$');
        expect(result.valid).toBe(true);
      });
    });

    describe('Patterns inválidos', () => {
      it('deve rejeitar pattern vazio', () => {
        const result = validateRegex('');
        expect(result.valid).toBe(false);
        expect(result.errorMessage).toContain('vazio');
      });

      it('deve rejeitar pattern nulo/undefined', () => {
        const result = validateRegex(null as unknown as string);
        expect(result.valid).toBe(false);
      });

      it('deve rejeitar pattern muito longo', () => {
        const longPattern = 'a'.repeat(MAX_PATTERN_LENGTH + 1);
        const result = validateRegex(longPattern);
        expect(result.valid).toBe(false);
        expect(result.errorMessage).toContain('muito longo');
      });

      it('deve rejeitar sintaxe inválida', () => {
        const result = validateRegex('[a-z');
        expect(result.valid).toBe(false);
        expect(result.errorMessage).toContain('Sintaxe');
      });
    });

    describe('Proteção ReDoS', () => {
      it('deve rejeitar (a+)+', () => {
        const result = validateRegex('(a+)+');
        expect(result.valid).toBe(false);
        expect(result.errorMessage).toContain('perigosa');
      });

      it('deve rejeitar (a*)*', () => {
        const result = validateRegex('(a*)*');
        expect(result.valid).toBe(false);
      });

      it('deve rejeitar (.*)+', () => {
        const result = validateRegex('(.*)+');
        expect(result.valid).toBe(false);
      });

      it('deve rejeitar (.+)+', () => {
        const result = validateRegex('(.+)+');
        expect(result.valid).toBe(false);
      });

      it('deve rejeitar ([a-zA-Z]+)*', () => {
        const result = validateRegex('([a-zA-Z]+)*');
        expect(result.valid).toBe(false);
      });

      it('deve rejeitar (a+)+$', () => {
        const result = validateRegex('(a+)+$');
        expect(result.valid).toBe(false);
      });

      it('deve rejeitar quantificadores aninhados dinâmicos', () => {
        const result = validateRegex('(x+y+)+');
        expect(result.valid).toBe(false);
      });
    });

    describe('Complexidade', () => {
      it('deve calcular complexidade para pattern simples', () => {
        const result = validateRegex('^[a-z]+$');
        expect(result.complexity).toBeGreaterThan(0);
        expect(result.complexity).toBeLessThan(50);
      });

      it('deve rejeitar pattern muito complexo', () => {
        const complex = '((a)(b)(c)(d)(e)(f)(g)(h)(i)(j)(k)+)+';
        const result = validateRegex(complex);
        expect(result.valid).toBe(false);
        expect(result.errorMessage).toContain('complexo');
      });
    });

    describe('Grupos de captura', () => {
      it('deve aceitar número razoável de grupos', () => {
        const result = validateRegex('(a)(b)(c)');
        expect(result.valid).toBe(true);
      });

      it('deve contar grupos corretamente', () => {
        const result = validateRegex('(?:a)(b)(?:c)(d)');
        expect(result.valid).toBe(true);
      });
    });
  });

  describe('getRegexValidationError', () => {
    it('deve retornar null para pattern válido', () => {
      expect(getRegexValidationError('^[a-z]+$')).toBeNull();
    });

    it('deve retornar mensagem para pattern inválido', () => {
      expect(getRegexValidationError('[a-z')).not.toBeNull();
    });

    it('deve retornar mensagem para pattern perigoso', () => {
      expect(getRegexValidationError('(a+)+')).not.toBeNull();
    });
  });

  describe('isRegexSafe', () => {
    it('deve retornar true para pattern seguro', () => {
      expect(isRegexSafe('^[a-z]+$')).toBe(true);
    });

    it('deve retornar false para pattern perigoso', () => {
      expect(isRegexSafe('(a+)+')).toBe(false);
    });

    it('deve retornar false para sintaxe inválida', () => {
      expect(isRegexSafe('[a-z')).toBe(false);
    });
  });

  describe('Casos de uso reais', () => {
    it('deve validar pattern de cartão de crédito', () => {
      const result = validateRegex('^\\d{4}[- ]?\\d{4}[- ]?\\d{4}[- ]?\\d{4}$');
      expect(result.valid).toBe(true);
    });

    it('deve validar pattern de data YYYYMMDD simplificado', () => {
      const result = validateRegex('^\\d{8}$');
      expect(result.valid).toBe(true);
    });

    it('deve validar pattern de hora HHMMSS simplificado', () => {
      const result = validateRegex('^\\d{6}$');
      expect(result.valid).toBe(true);
    });

    it('deve validar pattern de MCC', () => {
      const result = validateRegex('^\\d{4}$');
      expect(result.valid).toBe(true);
    });

    it('deve validar pattern de país ISO', () => {
      const result = validateRegex('^[A-Z]{2,3}$');
      expect(result.valid).toBe(true);
    });
  });
});

File: backend/src/test/java/com/rulex/util/RegexValidatorTest.java
package com.rulex.util;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.regex.Pattern;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

/** Testes para o RegexValidator com proteção contra ReDoS. */
class RegexValidatorTest {

  @Nested
  @DisplayName("Validação de Patterns")
  class ValidationTests {

    @Test
    @DisplayName("Deve aceitar pattern válido simples")
    void shouldAcceptSimpleValidPattern() {
      var result = RegexValidator.validate("^[a-z]+$");
      assertThat(result.valid()).isTrue();
      assertThat(result.errorMessage()).isNull();
    }

    @Test
    @DisplayName("Deve aceitar pattern de email")
    void shouldAcceptEmailPattern() {
      var result = RegexValidator.validate("^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}$");
      assertThat(result.valid()).isTrue();
    }

    @Test
    @DisplayName("Deve aceitar pattern de telefone")
    void shouldAcceptPhonePattern() {
      var result = RegexValidator.validate("^\\+?[0-9]{10,14}$");
      assertThat(result.valid()).isTrue();
    }

    @Test
    @DisplayName("Deve rejeitar pattern nulo")
    void shouldRejectNullPattern() {
      var result = RegexValidator.validate(null);
      assertThat(result.valid()).isFalse();
      assertThat(result.errorMessage()).contains("vazio");
    }

    @Test
    @DisplayName("Deve rejeitar pattern vazio")
    void shouldRejectEmptyPattern() {
      var result = RegexValidator.validate("");
      assertThat(result.valid()).isFalse();
      assertThat(result.errorMessage()).contains("vazio");
    }

    @Test
    @DisplayName("Deve rejeitar pattern muito longo")
    void shouldRejectTooLongPattern() {
      String longPattern = "a".repeat(RegexValidator.MAX_PATTERN_LENGTH + 1);
      var result = RegexValidator.validate(longPattern);
      assertThat(result.valid()).isFalse();
      assertThat(result.errorMessage()).contains("muito longo");
    }

    @Test
    @DisplayName("Deve rejeitar pattern com sintaxe inválida")
    void shouldRejectInvalidSyntax() {
      var result = RegexValidator.validate("[a-z");
      assertThat(result.valid()).isFalse();
      assertThat(result.errorMessage()).contains("Sintaxe");
    }
  }

  @Nested
  @DisplayName("Proteção contra ReDoS")
  class ReDoSProtectionTests {

    @Test
    @DisplayName("Deve rejeitar (a+)+ - catastrophic backtracking")
    void shouldRejectNestedQuantifier1() {
      var result = RegexValidator.validate("(a+)+");
      assertThat(result.valid()).isFalse();
      assertThat(result.errorMessage()).contains("perigosa");
    }

    @Test
    @DisplayName("Deve rejeitar (a*)*")
    void shouldRejectNestedQuantifier2() {
      var result = RegexValidator.validate("(a*)*");
      assertThat(result.valid()).isFalse();
    }

    @Test
    @DisplayName("Deve rejeitar (.*)+")
    void shouldRejectDotStarPlus() {
      var result = RegexValidator.validate("(.*)+");
      assertThat(result.valid()).isFalse();
    }

    @Test
    @DisplayName("Deve rejeitar (.+)+")
    void shouldRejectDotPlusPlus() {
      var result = RegexValidator.validate("(.+)+");
      assertThat(result.valid()).isFalse();
    }

    @Test
    @DisplayName("Deve rejeitar ([a-zA-Z]+)*")
    void shouldRejectCharClassQuantifier() {
      var result = RegexValidator.validate("([a-zA-Z]+)*");
      assertThat(result.valid()).isFalse();
    }

    @Test
    @DisplayName("Deve rejeitar (a+)+$")
    void shouldRejectEvilRegex() {
      var result = RegexValidator.validate("(a+)+$");
      assertThat(result.valid()).isFalse();
    }

    @Test
    @DisplayName("Deve rejeitar pattern muito complexo")
    void shouldRejectTooComplexPattern() {
      // Pattern com muitos grupos e quantificadores
      String complex = "((a)(b)(c)(d)(e)(f)(g)(h)(i)(j)(k)+)+";
      var result = RegexValidator.validate(complex);
      assertThat(result.valid()).isFalse();
    }
  }

  @Nested
  @DisplayName("Execução com Timeout")
  class TimeoutTests {

    @Test
    @DisplayName("Deve executar match simples com sucesso")
    void shouldMatchSimplePattern() throws RegexValidator.TimeoutException {
      Pattern pattern = Pattern.compile("^[a-z]+$");
      boolean result = RegexValidator.matchWithTimeout(pattern, "hello");
      assertThat(result).isTrue();
    }

    @Test
    @DisplayName("Deve retornar false para não-match")
    void shouldReturnFalseForNoMatch() throws RegexValidator.TimeoutException {
      Pattern pattern = Pattern.compile("^[a-z]+$");
      boolean result = RegexValidator.matchWithTimeout(pattern, "HELLO");
      assertThat(result).isFalse();
    }

    @Test
    @DisplayName("Deve retornar false para input nulo")
    void shouldReturnFalseForNullInput() throws RegexValidator.TimeoutException {
      Pattern pattern = Pattern.compile("^[a-z]+$");
      boolean result = RegexValidator.matchWithTimeout(pattern, null);
      assertThat(result).isFalse();
    }

    @Test
    @DisplayName("Deve truncar input muito longo")
    void shouldTruncateLongInput() throws RegexValidator.TimeoutException {
      Pattern pattern = Pattern.compile("^a+$");
      String longInput = "a".repeat(RegexValidator.MAX_INPUT_LENGTH + 1000);
      // Não deve lançar exceção, deve truncar
      boolean result = RegexValidator.matchWithTimeout(pattern, longInput);
      assertThat(result).isTrue(); // Truncado ainda é só 'a's
    }
  }

  @Nested
  @DisplayName("Safe Compile")
  class SafeCompileTests {

    @Test
    @DisplayName("Deve compilar pattern válido")
    void shouldCompileValidPattern() {
      Pattern pattern = RegexValidator.safeCompile("^[0-9]+$");
      assertThat(pattern).isNotNull();
      assertThat(pattern.matcher("12345").matches()).isTrue();
    }

    @Test
    @DisplayName("Deve retornar null para pattern perigoso")
    void shouldReturnNullForDangerousPattern() {
      Pattern pattern = RegexValidator.safeCompile("(a+)+");
      assertThat(pattern).isNull();
    }

    @Test
    @DisplayName("Deve retornar null para pattern inválido")
    void shouldReturnNullForInvalidPattern() {
      Pattern pattern = RegexValidator.safeCompile("[invalid");
      assertThat(pattern).isNull();
    }
  }

  @Nested
  @DisplayName("Casos de Uso Reais")
  class RealWorldTests {

    @Test
    @DisplayName("Deve validar pattern de CPF")
    void shouldValidateCpfPattern() {
      var result = RegexValidator.validate("^\\d{3}\\.\\d{3}\\.\\d{3}-\\d{2}$");
      assertThat(result.valid()).isTrue();
    }

    @Test
    @DisplayName("Deve validar pattern de CNPJ")
    void shouldValidateCnpjPattern() {
      var result = RegexValidator.validate("^\\d{2}\\.\\d{3}\\.\\d{3}/\\d{4}-\\d{2}$");
      assertThat(result.valid()).isTrue();
    }

    @Test
    @DisplayName("Deve validar pattern de CEP")
    void shouldValidateCepPattern() {
      var result = RegexValidator.validate("^\\d{5}-?\\d{3}$");
      assertThat(result.valid()).isTrue();
    }

    @Test
    @DisplayName("Deve validar pattern de cartão de crédito")
    void shouldValidateCreditCardPattern() {
      var result = RegexValidator.validate("^\\d{4}[- ]?\\d{4}[- ]?\\d{4}[- ]?\\d{4}$");
      assertThat(result.valid()).isTrue();
    }

    @Test
    @DisplayName("Deve validar pattern de UUID simplificado")
    void shouldValidateUuidPattern() {
      // Pattern simplificado para UUID (menos grupos)
      var result = RegexValidator.validate("^[0-9a-f-]{36}$");
      assertThat(result.valid()).isTrue();
    }

    @Test
    @DisplayName("Deve validar pattern de IP simplificado")
    void shouldValidateIpPattern() {
      // Pattern simplificado para IP (menos alternações)
      var result = RegexValidator.validate("^\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}$");
      assertThat(result.valid()).isTrue();
    }
  }
}
